### PR TITLE
Refactor the middleware and allow thread safety.

### PR DESCRIPTION
**Pull request by ruyrocha**

### CODE REVIEW

File: lib/wafris/middleware.rb [hunk 1]
```diff
@@ -7,6 +7,10 @@ def initialize(app)
     end
 
     def call(env)
+      dup.call!(env)
+    end
+
+    def call!(env)
```
Comment (ruyrocha) [inline lib/wafris/middleware.rb:13]: I'm redefining `call` method to allow thread safety.

File: lib/wafris/middleware.rb [hunk 2]
```diff
@@ -24,26 +28,38 @@ def call(env)
 
       Rack::Request.ip_filter = lambda { |ip| trusted_proxies.match?(ip) }
 
-      request = Rack::Request.new(env)
+      @env = env
+      @request = Rack::Request.new(env)
+
+      return forbidden! unless allowed_request?
```
Comment (ruyrocha) [inline lib/wafris/middleware.rb:34]: It's needed to halt the request earlier if the IP address is blocked.

File: lib/wafris/middleware.rb [hunk 3]
```diff
@@ -24,26 +28,38 @@ def call(env)
...
-        [403, {}, ['Blocked']]
-      end
+    def allowed_request?
+      Wafris.allow_request?(request)
     rescue Redis::TimeoutError
       LogSuppressor.puts_log(
         "[Wafris] Wafris timed out during processing. Request passed without rules check."
       )
-      @app.call(env)
+      true
```
Comment (ruyrocha) [inline lib/wafris/middleware.rb:57]: This is required as the intention is to continue with the request when there are exceptions.

**Comment by ruyrocha** [comment]
Closing in favor of major upcoming refactor.

**Comment by rmcastil** [comment]
@ruyrocha thanks for all the PRs. Sorry that I haven't responded sooner. I've been handling a major upgrade to Hub. What upcoming refactor do you have in mind?

**Comment by ruyrocha** [comment]
@rmcastil there are a few items to cover:

- it seems that the middleware is sharing too much state
- Hiredis can be used instead of the Ruby driver
- the logger is confusing 
- there is no need to pass the whole request object for `self.allow_request?`
- I'm in doubt about calling `dup` on the call method due to the way the Redis connection is defined

**Pull request closed without merging**